src/DependencyInjection/Compiler/ModelManagerCompilerPass.php

```diff
@@ -26,22 +27,40 @@
  */
 final class ModelManagerCompilerPass implements CompilerPassInterface
 {
+    public const MANAGER_TAG = 'sonata.admin.manager';
+
     public function process(ContainerBuilder $container): void
     {
         $availableManagers = [];
 
+        // NEXT_MAJOR: Replace the `foreach()` clause with the following one.
+        // foreach ($container->findTaggedServiceIds(self::MANAGER_TAG) as $id => $tags) {
         foreach ($container->getServiceIds() as $id) {
-            if (0 !== strpos($id, 'sonata.admin.manager.') || !is_subclass_of($container->getDefinition($id)->getClass(), ModelManagerInterface::class)) {
+            $definition = $container->findDefinition($id);
+
+            // NEXT_MAJOR: Remove this check.
+            if (!$definition->hasTag(self::MANAGER_TAG) && 0 !== strpos($id, 'sonata.admin.manager.')) {
                 continue;
             }
 
-            $availableManagers[$id] = $container->getDefinition($id);
+            // NEXT_MAJOR: Remove this check.
+            if (!$definition->hasTag(self::MANAGER_TAG)) {
+                @trigger_error(sprintf('Using the "%s" tag in the service "%s" is required since sonata-project/admin-bundle 3.x.', self::MANAGER_TAG, $id), E_USER_DEPRECATED);
```

Comment:
Technically, it is not required.
```suggestion
                @trigger_error(sprintf(
                    'Not setting the "%s" tag on the "%s" service is deprecated since sonata-project/admin-bundle 3.x.', self::MANAGER_TAG,
                    $id
                ), E_USER_DEPRECATED);
```